File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        Set<Integer>[] dp = new Set[9];
        
        for (int i=1;i<9;i++) {
            dp[i] = new HashSet<>();
            
            StringBuffer temp = new StringBuffer();
            int cnt = i;
            while (cnt-- > 0) temp.append(String.valueOf(N));
            
            dp[i].add(Integer.parseInt(temp.toString()));
            
            for (int j=1;j<=i/2;j++) {
                for (int n1: dp[j]) {
                    for (int n2: dp[i-j]) {
                        dp[i].add(n1+n2);
                        dp[i].add(n1*n2);
                        dp[i].add(n1-n2);
                        dp[i].add(n2-n1);
                        if (n1 > 0) dp[i].add(n2/n1);
                        if (n2 > 0) dp[i].add(n1/n2);
                    }
                }
            }
        }
        
        for (int i=1;i<9;i++) {
            for (int n: dp[i]) {
                if (n == number) return i;
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.py
# N은 1~8번 사용 가능
# 더하기, 빼기, 곱하기, 나누기, NNNNN ...

def solution(N, number):
    dp = [[] for _ in range(9)]
    for i in range(1, 9):
        dp[i].append(int(str(N)*i))
        if dp[i][0] == number: return i
        
        for j in range(1, i):
            
            for n1 in dp[j]:
                for n2 in dp[i-j]:
                    temp = [n1 + n2, n1 - n2, n1 * n2]
                    if n2 != 0: temp.append(n1 // n2)
                    for target in temp:
                        if target == number: return i
                    dp[i] += temp

    return -1